File: Books/A_Comprehensive_Approach_to_ML_and_AI/matlab/hyperparam_tutorial.m
close all; clear all; clc;

% Generate synthetic data for regression
n = 200;
X = linspace(0,10,n)';
y = sin(X) + 0.5*randn(n,1);

% Split data into training and test (holdout set)
testRatio = 0.2;
[X_train, y_train, X_test, y_test] = holdoutSplit(X, y, testRatio);

% Define hyperparameter grid for a simple ridge regression model (lambda)
lambdaGrid = logspace(-4, 2, 10);

% Perform 5-fold cross-validation to select the best lambda
k = 5;
cvErrors = zeros(length(lambdaGrid),1);
for i = 1:length(lambdaGrid)
    lambda = lambdaGrid(i);
    cvErrors(i) = kFoldCV(X_train, y_train, k, lambda);
end
[~, bestIdx] = min(cvErrors);
bestLambda = lambdaGrid(bestIdx);
fprintf('Best lambda from grid search: %.4f\n', bestLambda);

% Train final model on full training set using bestLambda
theta = ridgeRegression(X_train, y_train, bestLambda);
y_pred = X_test * theta;
testError = mean((y_test - y_pred).^2);
fprintf('Test MSE: %.4f\n', testError);

% Plot validation curve
figure;
semilogx(lambdaGrid, cvErrors, 'bo-', 'LineWidth', 2);
xlabel('Lambda');
ylabel('Cross-Validation MSE');
title('Validation Curve');
grid on;

% Compute and plot learning curves
[trainErrors, valErrors] = learningCurves(X_train, y_train, k, bestLambda);
figure;
plot(1:length(trainErrors), trainErrors, 'b-', 'LineWidth', 2); hold on;
plot(1:length(valErrors), valErrors, 'r-', 'LineWidth', 2);
xlabel('Training set size');
ylabel('MSE');
title('Learning Curves');
legend('Training Error', 'Validation Error');
grid on;

% save_all_figs_OPTION('results/hyperparam','png',1)

%% Function: Holdout Split
function [X_train, y_train, X_test, y_test] = holdoutSplit(X, y, testRatio)
    n = length(y);
    idx = randperm(n);
    nTest = round(testRatio * n);
    testIdx = idx(1:nTest);
    trainIdx = idx(nTest+1:end);
    X_train = X(trainIdx,:);
    y_train = y(trainIdx,:);
    X_test = X(testIdx,:);
    y_test = y(testIdx,:);
end

%% Function: k-Fold Cross-Validation for Ridge Regression
function cvError = kFoldCV(X, y, k, lambda)
    n = length(y);
    indices = crossvalind('Kfold', n, k);
    errors = zeros(k,1);
    for i = 1:k
        testIdx = (indices == i);
        trainIdx = ~testIdx;
        X_train = X(trainIdx,:);
        y_train = y(trainIdx,:);
        X_val = X(testIdx,:);
        y_val = y(testIdx,:);
        theta = ridgeRegression(X_train, y_train, lambda);
        y_pred = X_val * theta;
        errors(i) = mean((y_val - y_pred).^2);
    end
    cvError = mean(errors);
end

%% Function: Ridge Regression (Closed-Form)
function theta = ridgeRegression(X, y, lambda)
    % Closed-form solution for ridge regression: theta = (X'X + lambda*I)^(-1) X'y
    [n, p] = size(X);
    I = eye(p);
    theta = (X'*X + lambda * I) \ (X'*y);
end

%% Function: Learning Curves
function [trainErrors, valErrors] = learningCurves(X, y, k, lambda)
    n = length(y);
    % Define a set of training sizes
    sizes = round(linspace(floor(0.1*n), n, 10));
    trainErrors = zeros(length(sizes),1);
    valErrors = zeros(length(sizes),1);
    indices = crossvalind('Kfold', n, k);
    for s = 1:length(sizes)
        sizeTrain = sizes(s);
        % Randomly sample sizeTrain examples for training and use remaining for validation
        idx = randperm(n);
        trainIdx = idx(1:sizeTrain);
        valIdx = idx(sizeTrain+1:end);
        X_train = X(trainIdx,:);
        y_train = y(trainIdx,:);
        X_val = X(valIdx,:);
        y_val = y(valIdx,:);
        theta = ridgeRegression(X_train, y_train, lambda);
        y_pred_train = X_train * theta;
        y_pred_val = X_val * theta;
        trainErrors(s) = mean((y_train - y_pred_train).^2);
        valErrors(s) = mean((y_val - y_pred_val).^2);
    end
end

function idx = crossvalind(method, n, k)
% crossvalind Generate cross validation indices for k-fold cross validation.
%
%   idx = crossvalind('Kfold', n, k)
%
%   Inputs:
%       method - Cross-validation method. Only 'Kfold' is supported.
%       n      - Total number of observations (scalar).
%       k      - Number of folds (scalar).
%
%   Output:
%       idx    - n-by-1 vector of fold assignments, where each element is an
%                integer between 1 and k.
%
%   Example:
%       idx = crossvalind('Kfold', 100, 5);
%
%   This function randomly permutes the indices from 1 to n and assigns them
%   to k folds as equally as possible.

    % Check method
    if ~strcmpi(method, 'Kfold')
        error('Unsupported cross-validation method: %s', method);
    end

    % Check that k is less than or equal to n
    if k > n
        error('Number of folds k (%d) cannot exceed number of observations n (%d).', k, n);
    end

    % Initialize the index vector
    idx = zeros(n, 1);

    % Randomly permute the indices 1:n
    perm = randperm(n);

    % Compute fold sizes. If n is not divisible by k, the first 'remainder' folds
    % get one extra sample.
    fold_sizes = repmat(floor(n/k), k, 1);
    remainder = mod(n, k);
    fold_sizes(1:remainder) = fold_sizes(1:remainder) + 1;

    % Assign fold numbers to the permuted indices.
    start_idx = 1;
    for fold = 1:k
        end_idx = start_idx + fold_sizes(fold) - 1;
        idx(perm(start_idx:end_idx)) = fold;
        start_idx = end_idx + 1;
    end
end

File: Books/A_Comprehensive_Approach_to_ML_and_AI/R/hyperparam_tutorial.R
# hyperparam_tutorial.R
# Hyperparameter Tuning Tutorial
#
# This script demonstrates hyperparameter tuning for a simple ridge regression model.
# It includes:
#   - A holdout split into training and test sets.
#   - 5-fold cross-validation over a grid of lambda values.
#   - Training the final model using the best lambda.
#   - Plotting the validation curve and learning curves.
#
rm(list = ls())
graphics.off()
set.seed(1)

## Generate synthetic data for regression
n <- 200
X <- matrix(linspace(0, 10, n), ncol = 1)
y <- sin(X) + 0.5 * rnorm(n)

## Holdout Split Function
holdoutSplit <- function(X, y, testRatio) {
  n <- length(y)
  idx <- sample(1:n)
  nTest <- round(testRatio * n)
  testIdx <- idx[1:nTest]
  trainIdx <- idx[(nTest + 1):n]
  list(X_train = X[trainIdx, , drop = FALSE],
       y_train = y[trainIdx, , drop = FALSE],
       X_test = X[testIdx, , drop = FALSE],
       y_test = y[testIdx, , drop = FALSE])
}

split_res <- holdoutSplit(X, y, 0.2)
X_train <- split_res$X_train; y_train <- split_res$y_train
X_test <- split_res$X_test; y_test <- split_res$y_test

## Define Hyperparameter Grid
lambdaGrid <- 10^seq(-4, 2, length.out = 10)

## k-Fold Cross-Validation Function for Ridge Regression
kFoldCV <- function(X, y, k, lambda) {
  n <- nrow(X)
  # Create fold assignments
  folds <- sample(rep(1:k, length.out = n))
  errors <- numeric(k)
  for (i in 1:k) {
    testIdx <- which(folds == i)
    trainIdx <- setdiff(1:n, testIdx)
    X_train_cv <- X[trainIdx, , drop = FALSE]
    y_train_cv <- y[trainIdx, , drop = FALSE]
    X_val_cv <- X[testIdx, , drop = FALSE]
    y_val_cv <- y[testIdx, , drop = FALSE]
    theta <- ridgeRegression(X_train_cv, y_train_cv, lambda)
    y_pred <- X_val_cv %*% theta
    errors[i] <- mean((y_val_cv - y_pred)^2)
  }
  mean(errors)
}

ridgeRegression <- function(X, y, lambda) {
  p <- ncol(X)
  theta <- solve(t(X) %*% X + lambda * diag(p), t(X) %*% y)
  theta
}

cvErrors <- sapply(lambdaGrid, function(lambda) kFoldCV(X_train, y_train, 5, lambda))
bestIdx <- which.min(cvErrors)
bestLambda <- lambdaGrid[bestIdx]
cat(sprintf("Best lambda from grid search: %.4f\n", bestLambda))

## Train final model using bestLambda
theta <- ridgeRegression(X_train, y_train, bestLambda)
y_pred <- X_test %*% theta
testError <- mean((y_test - y_pred)^2)
cat(sprintf("Test MSE: %.4f\n", testError))

## Plot Validation Curve
par(mfrow = c(1,1))
plot(lambdaGrid, cvErrors, type = "b", log = "x", col = "blue", lwd = 2,
     xlab = "Lambda", ylab = "Cross-Validation MSE", main = "Validation Curve")
grid()

## Learning Curves Function
learningCurves <- function(X, y, k, lambda) {
  n <- nrow(X)
  sizes <- round(seq(0.1 * n, n, length.out = 10))
  trainErrors <- numeric(length(sizes))
  valErrors <- numeric(length(sizes))
  for (i in 1:length(sizes)) {
    sizeTrain <- sizes[i]
    idx <- sample(1:n)
    trainIdx <- idx[1:sizeTrain]
    valIdx <- idx[(sizeTrain + 1):n]
    X_train_subset <- X[trainIdx, , drop = FALSE]
    y_train_subset <- y[trainIdx, , drop = FALSE]
    X_val <- X[valIdx, , drop = FALSE]
    y_val <- y[valIdx, , drop = FALSE]
    theta <- ridgeRegression(X_train_subset, y_train_subset, lambda)
    y_pred_train <- X_train_subset %*% theta
    y_pred_val <- X_val %*% theta
    trainErrors[i] <- mean((y_train_subset - y_pred_train)^2)
    valErrors[i] <- mean((y_val - y_pred_val)^2)
  }
  list(trainErrors = trainErrors, valErrors = valErrors)
}

lc <- learningCurves(X_train, y_train, 5, bestLambda)
par(mfrow = c(1,1))
plot(lc$trainErrors, type = "l", col = "blue", lwd = 2,
     xlab = "Training set size index", ylab = "MSE", main = "Learning Curves")
lines(lc$valErrors, col = "red", lwd = 2)
legend("topright", legend = c("Training Error", "Validation Error"),
       col = c("blue", "red"), lwd = 2)
grid()
